File: tests/validate_scenarios.mjs
import jsyaml from 'js-yaml';
import fetch from 'node-fetch';
import yargs from 'yargs';
import { hideBin } from 'yargs/helpers';
import fs from 'fs/promises';

async function loadAndValidateYaml(gistUrl, filePath) {
    try {
        let yamlText;

        // Load YAML from local file or URL
        if (filePath) {
            // Read the YAML file from the provided file path
            yamlText = await fs.readFile(filePath, 'utf8');
        } else {
            // Fetch the YAML file from the provided Gist URL
            const response = await fetch(gistUrl);
            if (!response.ok) throw new Error('Failed to load data from Gist');
            yamlText = await response.text();
        }

        const data = jsyaml.load(yamlText);

        // Validate the YAML structure
        if (!Array.isArray(data.inputs) || !Array.isArray(data.calculations)) {
            throw new Error('Invalid YAML structure. Expected "inputs" and "calculations" fields as arrays.');
        }

        // Convert inputs array to a map for easier lookup
        const inputsMap = data.inputs.reduce((acc, input) => {
            if (!input.variable || typeof input.value === 'undefined') {
                throw new Error(`Invalid input entry: ${input.variable || 'Unknown Variable'}`);
            }
            acc[input.variable] = input;
            return acc;
        }, {});

        // Validate each calculation for required fields and run the equations
        data.calculations.forEach(calculation => {
            if (!calculation.title || !calculation.explanation || !Array.isArray(calculation.inputs)) {
                throw new Error(`Invalid calculation entry: ${calculation.title || 'Unknown Title'}`);
            }

            // Prepare a context for the equation with the available inputs
            const inputs = calculation.inputs.reduce((acc, key) => {
                if (!inputsMap[key] || typeof inputsMap[key].value === 'undefined') {
                    throw new Error(`Missing input value for key: ${key} in calculation: ${calculation.title}`);
                }
                acc[key] = inputsMap[key].value;
                return acc;
            }, {});

            // Attempt to evaluate the equation
            try {
                const result = new Function(...Object.keys(inputs), `return ${calculation.explanation}`)(...Object.values(inputs));
                console.log(`Calculation "${calculation.title}" evaluated successfully with result: ${result}`);
            } catch (error) {
                throw new Error(`Error evaluating calculation "${calculation.title}": ${error.message}`);
            }
        });

        console.log('All calculations evaluated successfully.');
    } catch (error) {
        console.error('Validation error:', error);
    }
}

// Using yargs to add optional CLI arguments for the URL or local file path
const argv = yargs(hideBin(process.argv))
    .option('url', {
        alias: 'u',
        type: 'string',
        description: 'URL of the YAML file to be validated',
        //default: 'https://gist.githubusercontent.com/nickmvincent/2c3e4ca38272b1d6b3041dd856e6cab7/raw/',
    })
    .option('file', {
        alias: 'f',
        type: 'string',
        description: 'Path to the local YAML file to be validated',
        default: 'data/data.yaml',

    })
    .help()
    .argv;

// Example usage
if (argv.file) {
    loadAndValidateYaml(null, argv.file);
} else {
    loadAndValidateYaml(argv.url, null);
}

File: tests/validate_scenarios.py
import argparse
import yaml
import requests
import os

# Load the YAML from a local file or a URL
def load_yaml(gist_url=None, file_path=None):
    try:
        if file_path:
            # Read YAML file from the provided local file path
            with open(file_path, 'r') as file:
                yaml_text = file.read()
        elif gist_url:
            # Fetch the YAML file from the provided Gist URL
            response = requests.get(gist_url)
            if response.status_code != 200:
                raise Exception("Failed to load data from Gist")
            yaml_text = response.text
        else:
            raise Exception("Either a file path or a URL must be provided")

        # Load the YAML data
        data = yaml.safe_load(yaml_text)

        return data

    except Exception as e:
        print(f"Error loading YAML: {e}")
        exit(1)

# Validate the input formatting and run the calculations
def validate_inputs(data):
    try:
        # Validate the YAML structure
        if not isinstance(data.get('inputs'), list) or not isinstance(data.get('calculations'), list):
            raise Exception('Invalid YAML structure. Expected "inputs" and "calculations" fields as lists.')

        # Convert inputs array to a dictionary for easier lookup
        inputs_map = {}
        for input_data in data['inputs']:
            if 'variable' not in input_data or 'value' not in input_data:
                raise Exception(f"Invalid input entry: {input_data.get('variable', 'Unknown Variable')}")
            inputs_map[input_data['variable']] = input_data

        # Validate each calculation for required fields and run the equations
        for calculation in data['calculations']:
            if 'title' not in calculation or 'explanation' not in calculation or not isinstance(calculation.get('inputs'), list):
                raise Exception(f"Invalid calculation entry: {calculation.get('title', 'Unknown Title')}")

            # Prepare a context for the equation with the available inputs
            inputs = {}
            for key in calculation['inputs']:
                if key not in inputs_map or inputs_map[key].get('value') is None:
                    raise Exception(f"Missing input value for key: {key} in calculation: {calculation['title']}")
                inputs[key] = inputs_map[key]['value']

            # Attempt to evaluate the equation
            try:
                # Use eval with caution. It's okay here because we control the source, but it can be dangerous.
                result = eval(calculation['explanation'], {}, inputs)
                print(f'Calculation "{calculation["title"]}" evaluated successfully with result: {result}')
            except Exception as error:
                raise Exception(f"Error evaluating calculation '{calculation['title']}': {error}")

        print('All calculations evaluated successfully.')

    except Exception as error:
        print(f'Validation error: {error}')

# Main function
if __name__ == "__main__":
    parser = argparse.ArgumentParser(description='Validate a YAML file from a URL or local file path.')
    parser.add_argument('--url', '-u', type=str, help='URL of the YAML file to be validated')
    parser.add_argument('--file', '-f', type=str, help='Path to the local YAML file to be validated', default='data/data.yaml')

    args = parser.parse_args()

    # Load and validate the YAML file
    yaml_data = None
    if args.file:
        yaml_data = load_yaml(file_path=args.file)
    elif args.url:
        yaml_data = load_yaml(gist_url=args.url)
    
    # Validate inputs and calculations if YAML was loaded
    if yaml_data:
        validate_inputs(yaml_data)
